xdsl/traits.py

```diff
@@ -68,6 +68,40 @@ def verify(self, op: Operation) -> None:
         raise VerifyException(f"'{op.name}' expects parent op to be one of {names}")
 
 
+@dataclass(frozen=True)
+class HasAncestor(OpTrait):
+    """
+    Constraint the operation to have a specific operation as ancestor, i.e. transitive
+    parent.
+    """
+
+    parameters: tuple[type[Operation], ...]
+
+    def __init__(self, *parameters: type[Operation]):
+        if not parameters:
+            raise ValueError("parameters must not be empty")
+        super().__init__(parameters)
```

Comment:
One way to enforce this syntactically is to have a first param, and remaining params
```suggestion
    def __init__(self, head_param: type[Operation], *tail_parameters: type[Operation]):
        super().__init__((head_param, *parameters))
```
Seems like the tuple construction cost is minimal compared to the runtime error annoyance, but that's kind of a matter of taste :)

Reply:
Right, will do 🙂